File: Алгоритмизация/Практики/Практическая №7/2/main.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int minBanknotes(int x, const std::vector<int>& coins) {
    std::vector<int> dp(x + 1, x + 2);
    dp[0] = 0;
    for (int i = 1; i <= x; ++i) {
        for (int coin : coins) {
            if (i >= coin) {
                dp[i] = std::min(dp[i], dp[i - coin] + 1);
            }
        }
    }
    return dp[x] > x + 1 ? -1 : dp[x];
}

int main() {
    int x;
    if (!(std::cin >> x)) return 0;
    int k;
    if (!(std::cin >> k)) return 0;
    std::vector<int> coins(k);
    for (int i = 0; i < k; ++i) {
        std::cin >> coins[i];
    }
    std::cout << minBanknotes(x, coins) << std::endl;
    return 0;
}

File: Алгоритмизация/Практики/Практическая №7/2/main.py
import sys

def min_banknotes(x: int, coins: list) -> int:
    dp = [x + 2] * (x + 1)
    dp[0] = 0
    for i in range(1, x + 1):
        for coin in coins:
            if i >= coin:
                dp[i] = min(dp[i], dp[i - coin] + 1)
    return -1 if dp[x] > x + 1 else dp[x]

if __name__ == "__main__":
    input_data = sys.stdin.read().split()
    if not input_data:
        sys.exit(0)
    x = int(input_data[0])
    k = int(input_data[1])
    coins = [int(i) for i in input_data[2:2+k]]
    print(min_banknotes(x, coins))
